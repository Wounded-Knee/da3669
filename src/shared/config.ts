// Webpack
export const WEBPACK_PORT = 3001;

// HTTP Server
export const HTTP_SERVER_PORT = 3000;

// WS Server
export const WS_SERVER_PORT = 3669;
export const WS_SERVER_HOST = 'localhost';

//Types
export const relationTypes = ['Reply', 'Stream'];
export const nodeTypes = ['Base', 'Document', 'User', 'Message', 'Economy'];

export const appName = 'D³';
export const headerText = ''; //appName || `merge℠`;

// Session
export const cookieName = `d3Session`;

// Node Server
export const subscriptionTimeoutMs = 1000 * 60 * 60; // 1 hour
