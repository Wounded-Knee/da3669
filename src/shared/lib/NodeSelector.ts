import { relationTypes, RelationType } from './RelationType';
import { NodeId } from '../all';
const debug = {
  addNodeIds: true,
};
const flatRelationTypes = relationTypes.flat(2).filter((relationType) => new RelationType(relationType).isPlural);

export interface INodeSelectorCfg {
  me: NodeId[];
  myRelations: {
    /*
     * null: Populate
     * true: hasRelation
     * false: lacksRelation
     */
    [key: string]: boolean | null;
  };
}

export class NodeSelector {
  cfg = <INodeSelectorCfg>{
    me: [],
    myRelations: {},
  };

  constructor(...nodeIds: NodeId[]) {
    return this.nodeIds(nodeIds);
  }

  serialize(): INodeSelectorCfg {
    return this.cfg;
  }

  deserialize(cfg: INodeSelectorCfg): NodeSelector {
    this.cfg = cfg;
    return this;
  }

  nodeId(nodeId: NodeId): NodeSelector {
    return this.nodeIds([nodeId]);
  }

  nodeIds(nodeIds: NodeId[]): NodeSelector {
    if (debug.addNodeIds) console.log('Adding Node IDs ', nodeIds);
    if (nodeIds.includes(undefined) || nodeIds.includes('undefined') || nodeIds === undefined) {
      debugger;
      console.error('Undefined Found');
    }
    this.cfg.me = [...this.cfg.me, ...nodeIds.filter((nodeId) => nodeId !== undefined)];
    return this;
  }

  populateRelation(...relationTypes: string[]): NodeSelector {
    return this.addRelations(null, relationTypes);
  }

  lacksRelation(...relationTypes: string[]): NodeSelector {
    return this.addRelations(false, relationTypes);
  }

  hasRelation(...relationTypes: string[]): NodeSelector {
    return this.addRelations(true, relationTypes);
  }

  addRelations(bool: boolean | null, relationTypes: string[]): NodeSelector {
    const theseRelationTypes = relationTypes.length ? relationTypes : flatRelationTypes;
    theseRelationTypes.forEach((relationType) => {
      this.cfg.myRelations[relationType] = bool;
    });
    return this;
  }

  equals(foreignSelector: INodeSelectorCfg | NodeSelector): boolean {
    if (foreignSelector instanceof NodeSelector) {
      return JSON.stringify(foreignSelector.serialize()) === JSON.stringify(this.serialize());
    } else {
      return JSON.stringify(foreignSelector) === JSON.stringify(this.serialize());
    }
  }
}
