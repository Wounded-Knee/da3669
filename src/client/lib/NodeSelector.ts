import { NodeSelector as SuperNodeSelector } from '../../shared/lib/NodeSelector';
import { RelationType } from '../../shared/lib/RelationType';
import { store } from './redux/store';
import { NodeId, INodeAll } from '../../shared/all';

export class NodeSelector extends SuperNodeSelector {
  async getMyNodes(): Promise<INodeAll[]> {
    return Promise.resolve(store.getState().nodes);
  }

  get nodes(): INodeAll[] {
    const { me, myRelations } = this.cfg;
    const allNodes = store.getState().nodes;
    const myNodeIds = me.map((_id) => _id.toString());
    const myNodes = allNodes.filter(({ _id }) => myNodeIds.length === 0 || myNodeIds.indexOf(_id.toString()) !== -1);

    return this.requiresRelations()
      ? allNodes.filter((node) => {
          const rv = Object.keys(myRelations).reduce((useThis: boolean, rel: string) => {
            if (myRelations[rel] === null) return useThis;

            // Combines all base node relations of [rel] type into one array
            const combinedBaseNodeRelations = myNodes.reduce((relations, myNode) => {
              return [
                ...relations,
                ...((myNode.rel && myNode.rel[new RelationType(rel).literal.plural]) || []).map((objectId) =>
                  objectId.toString(),
                ),
              ];
            }, []);
            const candidateNodeRelations = ((node.rel && node.rel[new RelationType(rel).literal.plural]) || []).map(
              (objectId) => objectId.toString(),
            );

            console.log(`My Real Relations (${rel})`, combinedBaseNodeRelations);
            console.log(
              `Candidate Relations (${new RelationType(rel).literal.plural})`,
              node.rel[new RelationType(rel).literal.plural] || [],
            );
            console.log('Base Node IDs ', me);
            console.log(
              'Candidate Relations / Base ID Intersection ',
              me.filter((myNodeId) => (node.rel[new RelationType(rel).literal.plural] || []).includes(myNodeId)),
            );

            return useThis || new RelationType(rel).isLiteral
              ? combinedBaseNodeRelations.includes(node._id)
              : !!me.filter((myNodeId) => candidateNodeRelations.includes(myNodeId.toString())).length;
          }, false);
          console.log('Node Included: ', rv, node, myRelations);
          return rv;
        })
      : myNodes;
  }
}

export const selectNodes = (...nodeIds: NodeId[]): NodeSelector => new NodeSelector(...nodeIds);
