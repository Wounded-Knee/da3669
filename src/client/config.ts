import { appName, WS_SERVER_HOST, WS_SERVER_PORT, initialState as sharedInitialState } from '../shared/config';
import { get } from './lib/LocalStorage';
import { Article, Chat as ChatIcon } from '@mui/icons-material';
import DocStore from './wireframes/docstore/DocStore';
import { Chat } from './wireframes/chat/Chat';

const localState = get(appName) || { ui: { drawers: {} } };

export const routes = [
  {
    route: '/chat',
    path: '/chat',
    icon: ChatIcon,
    text: 'Chat',
    component: Chat,
  },
  {
    path: '/docstore/:nodeId',
    component: DocStore,
  },
  {
    route: '/docstore',
    path: '/docstore',
    icon: Article,
    text: 'Doc Store',
    component: DocStore,
  },
];

export const initialState = {
  ...sharedInitialState,
  nodes: [],
  documents: [],
  user: {
    id: null,
  },
  ui: {
    ...localState.ui,
    drawers: {
      info: false,
      data: false,
      ...localState.ui.drawers,
    },
    ready: {
      webSocket: false,
    },
    docStore: {
      currentDoc: {},
    },
    selectedEntityIndex: null,
    selectedEntityHistory: [],
  },
};

export { appName, WS_SERVER_PORT, WS_SERVER_HOST };
