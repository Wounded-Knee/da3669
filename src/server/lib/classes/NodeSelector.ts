import { relationTypes, RelationType } from '../../../shared/lib/RelationType';
import { defaultModel } from '../nodes/all';
import { NodeSelector as SuperNodeSelector } from '../../../shared/lib/NodeSelector';
import { inspect } from 'util';
import { INodeAll } from '../../../shared/all';
import mongoose from 'mongoose';

const debug = {
  getNodes: true,
  filterMatchingNodes: false,
};

const {
  Types: { ObjectId },
} = mongoose;

export class NodeSelector extends SuperNodeSelector {
  get query(): any {
    const {
      cfg: { myRelations, me },
    } = this;

    const matches = {
      ...(me.length ? { _id: { $in: me.map((id) => new ObjectId(id)) } } : {}),
      ...Object.keys(myRelations).reduce((match, rel) => {
        return myRelations[rel] !== null
          ? {
              ...match,
              [`rel.${new RelationType(rel).literal.plural}`]: (() => {
                if (myRelations[rel] === true) {
                  return { $not: { $size: 0 } };
                } else if (myRelations[rel] === false) {
                  return { $size: 0 };
                } else {
                  return {};
                }
              })(),
            }
          : {};
      }, {}),
    };

    return [
      ...(Object.keys(matches).length
        ? [
            {
              $match: matches,
            },
          ]
        : []),
      ...Object.keys(myRelations).reduce((lookups, rel) => {
        if (myRelations[rel] === null) {
          return new RelationType(rel).isLiteral
            ? [
                ...lookups,
                {
                  $lookup: {
                    from: 'bases',
                    localField: `rel.${new RelationType(rel).literal.plural}`,
                    foreignField: '_id',
                    as: `rel.${new RelationType(rel).literal.plural}`,
                  },
                },
              ]
            : [
                ...lookups,
                {
                  $lookup: {
                    from: 'bases',
                    localField: '_id',
                    foreignField: `rel.${new RelationType(rel).literal.plural}`,
                    as: `rel.${new RelationType(rel).virtual.plural}`,
                  },
                },
              ];
        } else {
          return [];
        }
      }, []),
    ];
  }

  async getNodes(): Promise<INodeAll[]> {
    const query = this.query;
    try {
      const nodes = await defaultModel.aggregate(query).exec();
      if (debug.getNodes) console.log('Nodes ', nodes, inspect(query, { depth: null }));
      return nodes;
    } catch (e) {
      console.error('Query Error ', inspect(query, { depth: null }), e);
      return [];
    }
  }

  async filterMatchingNodes(nodeArray: INodeAll[]): Promise<INodeAll[]> {
    const {
      cfg: { myRelations, me },
    } = this;

    if (!defaultModel) {
      console.error('DefaultModel Problem ', defaultModel);
      return [];
    }

    const myNodes = await Promise.all(
      me.map(async (myNodeId) => {
        try {
          const node = await defaultModel.findById(myNodeId).exec();
          if (debug.filterMatchingNodes) console.log('Map ', myNodeId, node);
          return node;
        } catch (e) {
          console.error('Problem finding myNodes ', myNodeId, me, e);
        }
      }),
    );
    if (debug.filterMatchingNodes) console.log('My Nodes ', myNodes);

    return nodeArray.filter((node) => {
      return Object.keys(myRelations).reduce((useThis: boolean, rel: string) => {
        if (myRelations[rel] === null) return useThis;

        const myRealRelations = myNodes.reduce((relations, myNode) => {
          return [...relations, ...(myNode.rel[new RelationType(rel).literal.plural] || [])];
        }, []);

        return useThis || new RelationType(rel).isLiteral
          ? myRealRelations.includes(node._id)
          : !!me.filter((value) => (node.rel[new RelationType(rel).literal.plural] || []).includes(value)).length;
      }, false);
    });
  }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const selectNodes = (...args: any): NodeSelector => new NodeSelector(...args);
