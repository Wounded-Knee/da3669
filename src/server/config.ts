import dotenv from 'dotenv';
import findUp from 'find-up';
import path from 'path';
import fs from 'fs';
import { WS_SERVER_PORT, WS_SERVER_HOST, HTTP_SERVER_PORT as _HTTP_SERVER_PORT } from '../shared/config';

const IS_DEV = process.env.NODE_ENV !== 'production';

if (IS_DEV) {
  dotenv.config({ path: findUp.sync('.env') });
}

const packageJsonPath = path.join(process.cwd(), 'package.json');
const rawPackageJson = fs.readFileSync(packageJsonPath).toString();
const PackageJson = JSON.parse(rawPackageJson);
const { version: VERSION } = PackageJson;

// HTTP Server
const HTTP_SERVER_PORT = parseInt(process.env.PORT) || _HTTP_SERVER_PORT;
const WEBPACK_PORT = 8085; // For dev environment only

export { IS_DEV, VERSION, WEBPACK_PORT, HTTP_SERVER_PORT, WS_SERVER_HOST, WS_SERVER_PORT };
