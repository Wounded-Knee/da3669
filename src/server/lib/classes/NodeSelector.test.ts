// @ts-nocheck
import { NodeSelector } from './NodeSelector';
import mongoose from 'mongoose';
import '../../config';
const { ObjectId } = mongoose.Types;
// eslint-disable-next-line @typescript-eslint/no-var-requires
const util = require('util');

beforeAll(
  async () =>
    await mongoose.connect(
      process.env.MONGO_URL,
      {
        useNewUrlParser: true,
        useUnifiedTopology: true,
      },
      (err) => {
        if (err) {
          console.error(err);
          process.exit(1);
        }
      },
    ),
);

afterAll((done) => {
  // Closing the DB connection allows Jest to exit successfully.
  mongoose.connection.close();
  done();
});

describe('Query Generation', () => {
  describe('Without base node', () => {
    describe('populateRelation()', () => {
      test('Single', () => {
        const ns = new NodeSelector().populateRelation('upstreams');
        expect(ns.query).toMatchSnapshot();
      });

      test('Multiple', () => {
        const ns = new NodeSelector().populateRelation('upstreams', 'authors');
        expect(ns.query).toMatchSnapshot();
      });

      test('All', () => {
        const ns = new NodeSelector().populateRelation();
        expect(ns.query).toMatchSnapshot();
      });
    });

    describe('hasRelation()', () => {
      test('Single', () => {
        const ns = new NodeSelector().hasRelation('upstreams');
        expect(ns.query).toMatchSnapshot();
      });

      test('Multiple', () => {
        const ns = new NodeSelector().hasRelation('upstreams', 'authors');
        expect(ns.query).toMatchSnapshot();
      });

      test('All', () => {
        const ns = new NodeSelector().hasRelation();
        expect(ns.query).toMatchSnapshot();
      });
    });

    describe('lacksRelation()', () => {
      test('Single', () => {
        const ns = new NodeSelector().lacksRelation('upstreams');
        expect(ns.query).toMatchSnapshot();
      });
    });
  });

  describe('With base node', () => {
    describe('populateRelation()', () => {
      test('Single', () => {
        const ns = new NodeSelector('61be22843bb15b2d14696c25').populateRelation('upstreams');
        expect(ns.query).toMatchSnapshot();
      });

      test('Multiple', () => {
        const ns = new NodeSelector('61be22843bb15b2d14696c25').populateRelation('upstreams', 'authors');
        expect(ns.query).toMatchSnapshot();
      });

      test('All', () => {
        const ns = new NodeSelector('61be22843bb15b2d14696c25').populateRelation();
        expect(ns.query).toMatchSnapshot();
      });
    });

    describe('hasRelation()', () => {
      test('Single', () => {
        const ns = new NodeSelector('61be22843bb15b2d14696c25').hasRelation('upstreams');
        expect(ns.query).toMatchSnapshot();
      });

      test('Multiple', () => {
        const ns = new NodeSelector('61be22843bb15b2d14696c25').hasRelation('upstreams', 'authors');
        expect(ns.query).toMatchSnapshot();
      });

      test('All', () => {
        const ns = new NodeSelector('61be22843bb15b2d14696c25').hasRelation();
        expect(ns.query).toMatchSnapshot();
      });
    });

    describe('lacksRelation()', () => {
      test('Single', () => {
        const ns = new NodeSelector('61be22843bb15b2d14696c25').lacksRelation('upstreams');
        expect(ns.query).toMatchSnapshot();
      });
    });
  });
});

describe('Node Matching', () => {
  test('Matching nodes are filtered', async () => {
    expect.assertions(0);
    const ns = new NodeSelector('61be23e15d55c9e5d68a2492').populateRelation('downstreams');
    const matchingNodes = await ns.filterMatchingNodes([
      {
        _id: new ObjectId('61be23585d55c9e5d68a247d'),
        text: 'Cookie',
        author: 'plugh',
        rel: {
          upstreams: [new ObjectId('61be23e15d55c9e5d68a2492')],
        },
      },
    ]);

    return matchingNodes;
  });
});
