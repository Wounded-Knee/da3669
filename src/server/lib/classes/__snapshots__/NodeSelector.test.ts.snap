// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Generation With base node hasRelation() All 1`] = `
Array [
  Object {
    "$match": Object {
      "_id": Object {
        "$in": Array [
          "61be22843bb15b2d14696c25",
        ],
      },
      "rel.authors": Object {
        "$not": Object {
          "$size": 0,
        },
      },
      "rel.children": Object {
        "$not": Object {
          "$size": 0,
        },
      },
      "rel.upstreams": Object {
        "$not": Object {
          "$size": 0,
        },
      },
    },
  },
]
`;

exports[`Query Generation With base node hasRelation() Multiple 1`] = `
Array [
  Object {
    "$match": Object {
      "_id": Object {
        "$in": Array [
          "61be22843bb15b2d14696c25",
        ],
      },
      "rel.authors": Object {
        "$not": Object {
          "$size": 0,
        },
      },
      "rel.upstreams": Object {
        "$not": Object {
          "$size": 0,
        },
      },
    },
  },
]
`;

exports[`Query Generation With base node hasRelation() Single 1`] = `
Array [
  Object {
    "$match": Object {
      "_id": Object {
        "$in": Array [
          "61be22843bb15b2d14696c25",
        ],
      },
      "rel.upstreams": Object {
        "$not": Object {
          "$size": 0,
        },
      },
    },
  },
]
`;

exports[`Query Generation With base node lacksRelation() Single 1`] = `
Array [
  Object {
    "$match": Object {
      "_id": Object {
        "$in": Array [
          "61be22843bb15b2d14696c25",
        ],
      },
      "rel.upstreams": Object {
        "$size": 0,
      },
    },
  },
]
`;

exports[`Query Generation With base node populateRelation() All 1`] = `
Array [
  Object {
    "$match": Object {
      "_id": Object {
        "$in": Array [
          "61be22843bb15b2d14696c25",
        ],
      },
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.upstreams",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.upstreams",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.downstreams",
      "foreignField": "rel.upstreams",
      "from": "bases",
      "localField": "_id",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.children",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.children",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.parents",
      "foreignField": "rel.children",
      "from": "bases",
      "localField": "_id",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.authors",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.authors",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.works",
      "foreignField": "rel.authors",
      "from": "bases",
      "localField": "_id",
    },
  },
]
`;

exports[`Query Generation With base node populateRelation() Multiple 1`] = `
Array [
  Object {
    "$match": Object {
      "_id": Object {
        "$in": Array [
          "61be22843bb15b2d14696c25",
        ],
      },
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.upstreams",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.upstreams",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.authors",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.authors",
    },
  },
]
`;

exports[`Query Generation With base node populateRelation() Single 1`] = `
Array [
  Object {
    "$match": Object {
      "_id": Object {
        "$in": Array [
          "61be22843bb15b2d14696c25",
        ],
      },
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.upstreams",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.upstreams",
    },
  },
]
`;

exports[`Query Generation Without base node hasRelation() All 1`] = `
Array [
  Object {
    "$match": Object {
      "rel.authors": Object {
        "$not": Object {
          "$size": 0,
        },
      },
      "rel.children": Object {
        "$not": Object {
          "$size": 0,
        },
      },
      "rel.upstreams": Object {
        "$not": Object {
          "$size": 0,
        },
      },
    },
  },
]
`;

exports[`Query Generation Without base node hasRelation() Multiple 1`] = `
Array [
  Object {
    "$match": Object {
      "rel.authors": Object {
        "$not": Object {
          "$size": 0,
        },
      },
      "rel.upstreams": Object {
        "$not": Object {
          "$size": 0,
        },
      },
    },
  },
]
`;

exports[`Query Generation Without base node hasRelation() Single 1`] = `
Array [
  Object {
    "$match": Object {
      "rel.upstreams": Object {
        "$not": Object {
          "$size": 0,
        },
      },
    },
  },
]
`;

exports[`Query Generation Without base node lacksRelation() Single 1`] = `
Array [
  Object {
    "$match": Object {
      "rel.upstreams": Object {
        "$size": 0,
      },
    },
  },
]
`;

exports[`Query Generation Without base node populateRelation() All 1`] = `
Array [
  Object {
    "$lookup": Object {
      "as": "rel.upstreams",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.upstreams",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.downstreams",
      "foreignField": "rel.upstreams",
      "from": "bases",
      "localField": "_id",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.children",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.children",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.parents",
      "foreignField": "rel.children",
      "from": "bases",
      "localField": "_id",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.authors",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.authors",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.works",
      "foreignField": "rel.authors",
      "from": "bases",
      "localField": "_id",
    },
  },
]
`;

exports[`Query Generation Without base node populateRelation() Multiple 1`] = `
Array [
  Object {
    "$lookup": Object {
      "as": "rel.upstreams",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.upstreams",
    },
  },
  Object {
    "$lookup": Object {
      "as": "rel.authors",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.authors",
    },
  },
]
`;

exports[`Query Generation Without base node populateRelation() Single 1`] = `
Array [
  Object {
    "$lookup": Object {
      "as": "rel.upstreams",
      "foreignField": "_id",
      "from": "bases",
      "localField": "rel.upstreams",
    },
  },
]
`;
